File: stack/84_largest_rectangle_in_histogram.cpp
#include <iostream>
#include <vector>
using namespace std;

/* Given an array of integers heights representing the histogram's bar height
where the width of each bar is 1, return the area of the largest rectangle in
the histogram.

https://leetcode.com/problems/largest-rectangle-in-histogram/
O(n), O(n) */

class Solution {
public:
    int largestRectangleArea(vector<int> &heights) {
        int res = 0;
        vector<pair<int, int>> s;
        int n = heights.size();
        for (int i = 0; i < n; i++) {
            int h = heights[i];
            int start = i;
            while (!s.empty() && s.back().second > h) {
                int index = s.back().first;
                int height = s.back().second;
                s.pop_back();
                res = max(res, height * (i - index));
                start = index;
            }
            s.push_back({start, h});
        }
        for (int i = 0; i < s.size(); i++) {
            res = max(res, s[i].second * (n - s[i].first));
        }
        return res;
    }
};

int main(int argc, char *argv[]) {
    vector<int> heights = {2, 1, 5, 6, 2, 3};
    cout << Solution().largestRectangleArea(heights) << endl;
    return 0;
}

File: stack/84_largest_rectangle_in_histogram.py
from typing import List

"""
Given an array of integers heights representing the histogram's bar height
where the width of each bar is 1, return the area of the largest rectangle in
the histogram.

https://leetcode.com/problems/largest-rectangle-in-histogram/
O(n), O(n)
"""


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        res = 0
        stack = []
        for i, h in enumerate(heights):
            start = i
            while stack and stack[-1][1] > h:
                index, height = stack.pop()
                res = max(res, height * (i - index))
                start = index
            stack.append((start, h))

        for i, h in stack:
            res = max(res, h * (len(heights) - i))
        return res


if __name__ == "__main__":
    heights = [2, 1, 5, 6, 2, 3]
    print(Solution().largestRectangleArea(heights))
